server/src/main/java/org/opensearch/cluster/routing/allocation/allocator/LocalShardsBalancer.java

```diff
@@ -600,13 +600,15 @@ void moveShards() {
                 final BalancedShardsAllocator.ModelNode sourceNode = nodes.get(shardRouting.currentNodeId());
                 final BalancedShardsAllocator.ModelNode targetNode = nodes.get(moveDecision.getTargetNode().getId());
                 sourceNode.removeShard(shardRouting);
+                --totalShardCount;
```

Comment:
Improvement is definitely good. I have just few concerns around maintaining inc/dec counts of total shards along with every remove/add shard operation. 
1. Can you please move the two operations (add/remove and inc/dec) to a single method since both are necessary now?  Something like [this](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/cluster/routing/RoutingNodes.java#L182).
2. Also, it looks like there is a need of an invariant assertion now to be invoked on start and end of the moveShards method. Ref [CP](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/index/seqno/ReplicationTracker.java#L621).

Reply:
Also, can you please check and confirm if tests are covering all add/remove shards scenarios?

Reply:
@vikasvb90 what would you assert on in moveShards? I agree we can add more unit test to cover possible cases for addShard/ removeShard combination

Reply:
This is to ensure tests breaks if a addShard or removeShard is added without changing totalShardCount.